subprojects/dependency-management/src/main/java/org/gradle/internal/component/external/model/AbstractMutableModuleComponentResolveMetadata.java

```diff
@@ -54,6 +59,8 @@
     private ImmutableList<? extends ModuleComponentArtifactMetadata> artifactOverrides;
     private ImmutableMap<String, T> configurations;
 
+    protected final Map<String, DependencyMetadataRules> dependencyMetadataRules = Maps.newHashMap();
```

Comment:
❌ Needs to be linked to guarantee order of rules.

Reply:
This maps _variant (or configuration)_ name to the list of rules per variant. For each variant a list is defined in `DependencyMetadataRules` (`List<Action<DependenciesMetadata>> actions`). Only one list is selected in the end, in which the order is preserved.

Reply:
Got it.